generator/src/lib.rs

```diff
@@ -90,11 +94,83 @@ pub fn derive_parser(input: TokenStream, include_grammar: bool) -> TokenStream {
         Err(error) => panic!("error parsing \n{}", error.renamed_rules(rename_meta_rule)),
     };
 
+    let grammar_doc = consume_grammar_doc(pairs.clone());
+    let line_docs = consume_line_docs(pairs.clone());
+
     let defaults = unwrap_or_report(validator::validate_pairs(pairs.clone()));
     let ast = unwrap_or_report(parser::consume_rules(pairs));
     let optimized = optimizer::optimize(ast);
 
-    generator::generate(name, &generics, path, optimized, defaults, include_grammar)
+    let doc_comment = &DocComment::new(grammar_doc, line_docs);
+
+    generator::generate(
+        name,
+        &generics,
+        path,
+        optimized,
+        defaults,
+        doc_comment,
+        include_grammar,
+    )
+}
+
+/// Consume grammar doc into String, multi-line joined with `\n`
+fn consume_grammar_doc(pairs: Pairs<'_, Rule>) -> String {
+    let mut docs = vec![];
+    for pair in pairs {
+        if pair.as_rule() == Rule::grammar_doc {
+            let inner_doc = pair.into_inner().next().unwrap();
+            docs.push(inner_doc.as_str());
+        }
+    }
+
+    docs.join("\n")
```

Comment:
this may potentially be replaced with string building, i.e. have an empty string, and then `docs.push_str(inner_doc.as_str())` and `docs.push('\n')`

Reply:
That will have a extra `\n` in tail.

Reply:
Or I use `push_str` and `pop` last char?

```
let out = String::new();

for pair in pairs {
  out.push_str(inner_doc.as_str())
  out.push('\n')
}

out.pop()
out
```

Reply:
I guess it'll need a check if non-empty, but it could work

Reply:
This part have updated.